### PR TITLE
Added logging per Brain of time to update policy, time elapsed during training, time to collect experiences, buffer length, average return

### PR DESCRIPTION
**Pull request by eshvk**

### CODE REVIEW

**Comment by eshvk** [comment]
@awjuliani and @vincentpierre: Please comment/review since @harper-u3d is out this week.

File: ml-agents/mlagents/trainers/trainer.py [hunk 1]
```diff
@@ -151,6 +161,34 @@ def export_model(self):
         """
         self.policy.export_model()
 
+    def write_training_metrics(self):
+        """
+        Write training metrics to a CSV  file
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/trainer.py]: Extra space on this line.

**Comment by awjuliani** [comment]
@eshvk  So I was wrong. It looks like you are taking the correct reward. I wonder why it doesn't necessarily track the reward printed during the update frequency then...

File: ml-agents/mlagents/trainers/trainer.py [hunk 2]
```diff
@@ -32,11 +33,18 @@ def __init__(self, brain, trainer_parameters, training, run_id):
         self.brain_name = brain.brain_name
         self.run_id = run_id
         self.trainer_parameters = trainer_parameters
+        self.time_start_experience_collection = None
+        self.delta_last_experience_collection = None
+        self.cumulative_returns_since_policy_update = [0]
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer.py]: Why is this a zero again ?

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 3]
```diff
@@ -338,7 +341,14 @@ def update_policy(self):
...
             self.stats['Losses/Forward Loss'].append(np.mean(forward_total))
             self.stats['Losses/Inverse Loss'].append(np.mean(inverse_total))
+        self.last_buffer_length = len(self.training_buffer.update_buffer['actions'])
         self.training_buffer.reset_update_buffer()
+        self.delta_last_experience_collection = time() - \
+                                                self.time_start_experience_collection
+        self.time_start_experience_collection = time()
+        self.last_mean_return = np.mean(self.
+                                        cumulative_returns_since_policy_update)
+        self.all_rewards_since_policy_update = [0]
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/ppo/trainer.py]: Why is this a zero?

File: ml-agents/mlagents/trainers/trainer.py [hunk 4]
```diff
@@ -151,6 +161,34 @@ def export_model(self):
...
+    def write_training_metrics(self):
+        """
+        Write training metrics to a CSV  file
+        :return:
+        """
+        log_file = self.summary_path + '.csv'
+        with open(log_file, 'w') as f:
+            f.writelines('%s\n' % l for l in self.training_metrics)
+
+    def write_training_metric(self, delta_update_policy,
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/trainer.py]: Some comment for this method.

File: ml-agents/mlagents/trainers/trainer.py (hunk 2 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: This was just intended as an explicit initialization value. Thinking about it, this creates a problem where the mean will lower than expected, will fix.

File: ml-agents/mlagents/trainers/trainer.py (hunk 4 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: fixed

**Comment by eshvk** [comment]
@vincentpierre check updates reg.  logging

**Comment by harperj** [comment]
Moved some fixes I'd added to the branch back over to `develop-jh-parallel-envs` to avoid confusion.

File: ml-agents/mlagents/trainers/bc/offline_trainer.py [hunk 5]
```diff
@@ -15,17 +15,19 @@
...
         :param  trainer_parameters: The parameters for the trainer (dictionary).
         :param training: Whether the trainer is set for training.
         :param load: Whether the model should be loaded.
         :param seed: The seed the model will be initialized with
         :param run_id: The The identifier of the current run
+        :param debug_flag: Log debug statements
         """
         super(OfflineBCTrainer, self).__init__(
-            brain, trainer_parameters, training, load, seed, run_id)
+            brain, trainer_parameters, training, load, seed, run_id, debug_flag)
```
Comment (harperj) [inline ml-agents/mlagents/trainers/bc/offline_trainer.py]: Any thoughts on why this should be a debug print statement whereas the logging every N steps is not?  Relevant to the whole discussion we had about whether to print at `summary_freq` at all.  Curious how you came to this decision.

File: ml-agents/mlagents/trainers/trainer.py [hunk 6]
```diff
@@ -20,22 +21,33 @@ class UnityTrainerException(UnityException):
...
+        self.cumulative_returns_since_policy_update = []
+        self.last_mean_return = None
+        self.last_buffer_length = 0
         self.is_training = training
         self.stats = {}
         self.summary_writer = None
+        self.training_metrics  = []
         self.policy = None
+        if debug_flag:
+            logger.setLevel('DEBUG')
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer.py]: Seems like an alternative to setting the logging level in multiple places would be to set the logging level at the entrypoint and have all of these modules use shared loggers (`trainer_logger`, `env_logger`, etc).

File: ml-agents/mlagents/trainers/trainer.py [hunk 7]
```diff
@@ -20,22 +21,33 @@ class UnityTrainerException(UnityException):
...
         :dict trainer_parameters: The parameters for the trainer (dictionary).
         :bool training: Whether the trainer is set for training.
         :int run_id: The identifier of the current run
+        :bool debug_flag: Checks debug_flag
         """
         self.param_keys = []
         self.brain_name = brain.brain_name
         self.run_id = run_id
         self.trainer_parameters = trainer_parameters
+        self.time_start_experience_collection = None
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer.py]: If these added fields are purely for logging rather than added trainer functionality it would probably make sense to keep them in a separate code path (e.g. a separate class)

File: ml-agents/mlagents/trainers/trainer.py [hunk 8]
```diff
@@ -151,19 +165,58 @@ def export_model(self):
...
+
+    def write_training_metric(self, delta_update_policy,
+                              delta_train_start):
+        """
+        Write training  metrics to be eventually be persisted in a  CSV
+        :param delta_update_policy: Time (in seconds) to update policy
+        :param delta_train_start: Time (in  seconds) since training started.
+        :return:
+        """
+        logger.debug("Time to update Policy: {:0.3f} s "
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer.py]: What do you think about adding newlines after each of these metrics (e.g. like in RLLib) and including a header with what training step they refer to?

File: ml-agents/mlagents/trainers/trainer.py [hunk 9]
```diff
@@ -151,19 +165,58 @@ def export_model(self):
...
+                    "Time elapsed since training: {:0.3f} s "
+                    "Time for experience collection: {:0.3f} s "
+                    "Buffer Length: {} "
+                    "Returns : {:0.3f}"
+                    .format(delta_update_policy, delta_train_start,
+                            self.delta_last_experience_collection,
+                            self.last_buffer_length,
+                            self.last_mean_return)
+                    )
+        training_metric ='{:.3f}, {:.3f}, {:.3f}, {}, {:.3f}'.format(
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer.py]: Might want to consider using the CSV module here: https://docs.python.org/3/library/csv.html.

The CSV should also have headers -- by adding headers we might be able to also reduce duplication between the logging above and the CSV output here.  Not 100% sure about this.

File: ml-agents/mlagents/trainers/trainer.py [hunk 10]
```diff
@@ -151,19 +165,58 @@ def export_model(self):
...
+        Write training  metrics to be eventually be persisted in a  CSV
+        :param delta_update_policy: Time (in seconds) to update policy
+        :param delta_train_start: Time (in  seconds) since training started.
+        :return:
+        """
+        logger.debug("Time to update Policy: {:0.3f} s "
+                    "Time elapsed since training: {:0.3f} s "
+                    "Time for experience collection: {:0.3f} s "
+                    "Buffer Length: {} "
+                    "Returns : {:0.3f}"
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer.py]: In my experience running early experiments, the returns shown here can be misleading when compared to rewards -- for example they sometimes start very high and then lower significantly over time whereas rewards are steadily increasing.  We might want to also include rewards here.

Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: Can you be a bit more specific in terms of which rewards you mean here? Do you mean cumulative rewards?

File: ml-agents/mlagents/trainers/bc/offline_trainer.py (hunk 5 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/bc/offline_trainer.py]: We discussed a bit more and decided to keep the current functionality as is and move this to be a debug print b/c it wasn't clear whether this was actually useful to the person monitoring the command line.

File: ml-agents/mlagents/trainers/trainer.py (hunk 9 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: Will do

File: ml-agents/mlagents/trainers/trainer.py (hunk 6 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: Good idea, will fix

File: ml-agents/mlagents/trainers/bc/offline_trainer.py (hunk 5 above)
Comment (harperj) [reply ml-agents/mlagents/trainers/bc/offline_trainer.py]: Sounds good

File: ml-agents/mlagents/trainers/trainer.py (hunk 6 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: PTAL

File: ml-agents/mlagents/trainers/trainer.py (hunk 7 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: PTAL

File: ml-agents/mlagents/trainers/trainer.py (hunk 8 above)
Comment (eshvk) [reply ml-agents/mlagents/trainers/trainer.py]: PTAL

**Comment by eshvk** [comment]
@harperj I am closing this and rebasing against develop. https://github.com/Unity-Technologies/ml-agents/pull/1858
